src/D2L.CodeStyle.Analyzers/Language/OnlyVisibleToAnalyzer.Model.cs

```diff
@@ -0,0 +1,123 @@
+﻿using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+
+namespace D2L.CodeStyle.Analyzers.Language {
+
+	public sealed partial class OnlyVisibleToAnalyzer {
+
+		internal readonly struct Model {
+
+			private const string OnlyVisibleToAttributeMetadataName = "D2L.CodeStyle.Annotations.Contract.OnlyVisibleToAttribute";
+
+			private readonly Compilation m_compilation;
+			private readonly INamedTypeSymbol m_onlyVisibleToAttribute;
+
+			private readonly ConcurrentDictionary<ISymbol, ImmutableHashSet<INamedTypeSymbol>?> m_visibilityCache =
+				new ConcurrentDictionary<ISymbol, ImmutableHashSet<INamedTypeSymbol>?>( SymbolEqualityComparer.Default );
+
+			private Model(
+					Compilation compilation,
+					INamedTypeSymbol onlyVisibleToAttribute
+				) {
+
+				m_compilation = compilation;
+				m_onlyVisibleToAttribute = onlyVisibleToAttribute;
+			}
+
+			public static Model? TryCreate( Compilation compilation ) {
+
+				INamedTypeSymbol? onlyVisibleToAttribute = compilation.GetTypeByMetadataName( OnlyVisibleToAttributeMetadataName );
+				if( onlyVisibleToAttribute == null ) {
+					return null;
+				}
+
+				return new Model( compilation, onlyVisibleToAttribute );
+			}
+
+			public bool IsVisibleTo( INamedTypeSymbol caller, ISymbol member ) {
+
+				ImmutableHashSet<INamedTypeSymbol>? restrictions = m_visibilityCache
+					.GetOrAdd( member, GetVisibilityRestrictions );
+	
+				if( restrictions == null ) {
+					return true;
+				}
+
+				if( restrictions.Contains( caller ) ) {
+					return true;
+				}
+
+				if( SymbolEqualityComparer.Default.Equals( caller, member.ContainingType ) ) {
+					return true;
+				}
+
+				return false;
+			}
+
+			private ImmutableHashSet<INamedTypeSymbol>? GetVisibilityRestrictions( ISymbol member ) {
+
+				ImmutableArray<AttributeData> attributes = member.GetAttributes();
+
+				int attributeCount = attributes.Length;
+				if( attributeCount == 0 ) {
+					return null;
+				}
+
+				ImmutableHashSet<INamedTypeSymbol>.Builder? restrictions = null;
+
+				for( int i = 0; i < attributeCount; i++ ) {
+					AttributeData attribute = attributes[ i ];

```

Comment:
Would prefer to stick to foreach unless we've an explicit rationale not to.

```suggestion
				foreach( AttributeData attribute in attributes ) {
```

Reply:
Done